File: stepik-java-api/src/main/java/org/stepik/api/actions/StepikFilesAction.java
package org.stepik.api.actions;

import org.apache.http.HttpHeaders;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.stepik.api.client.ClientResponse;
import org.stepik.api.client.StatusCodes;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.client.TransportClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.exceptions.StepikUnauthorizedException;
import org.stepik.api.objects.auth.TokenInfo;

import java.util.HashMap;
import java.util.Map;

/**
 * @author meanmail
 */
public class StepikFilesAction extends StepikAbstractAction {
    private static final Logger logger = LoggerFactory.getLogger(StepikFilesAction.class);

    public StepikFilesAction(@NotNull StepikApiClient stepikApiClient) {
        super(stepikApiClient);
    }

    @NotNull
    public String get(@NotNull String url, String contentType) {
        StepikApiClient stepikApi = getStepikApiClient();
        TransportClient transportClient = stepikApi.getTransportClient();

        Map<String, String> headers = new HashMap<>();
        TokenInfo tokenInfo = stepikApi.getTokenInfo();
        String accessToken = tokenInfo.getAccessToken();
        if (accessToken != null) {
            String tokenType = tokenInfo.getTokenType();
            headers.put(HttpHeaders.AUTHORIZATION, tokenType + " " + accessToken);
        }
        headers.put(HttpHeaders.CONTENT_TYPE, contentType);

        ClientResponse response = transportClient.get(stepikApi, url, headers);

        if (response.getStatusCode() / 100 != 2) {
            String message = "Failed query to " + url + " returned the status code " + response.getStatusCode();
            logger.warn(message);

            if (response.getStatusCode() == StatusCodes.SC_UNAUTHORIZED) {
                throw new StepikUnauthorizedException(message);
            } else {
                throw new StepikClientException(message);
            }
        }

        String result = response.getBody();

        if (result == null) {
            throw new StepikClientException("Request successfully but the response body is null: " + url);
        }
        return result;
    }
}

File: stepik-java-api/src/main/kotlin/org/stepik/api/actions/StepikFilesAction.kt
package org.stepik.api.actions

import org.apache.http.HttpHeaders
import org.slf4j.LoggerFactory
import org.stepik.api.client.StatusCodes
import org.stepik.api.client.StepikApiClient
import org.stepik.api.exceptions.StepikClientException
import org.stepik.api.exceptions.StepikUnauthorizedException
import java.util.*

class StepikFilesAction(stepikApiClient: StepikApiClient) : StepikAbstractAction(stepikApiClient) {
    
    operator fun get(url: String, contentType: String): String {
        val stepikApi = stepikApiClient
        val transportClient = stepikApi.transportClient
        
        val headers = HashMap<String, String>()
        val tokenInfo = stepikApi.tokenInfo
        val accessToken = tokenInfo?.accessToken
        if (accessToken != null) {
            val tokenType = tokenInfo.tokenType
            headers[HttpHeaders.AUTHORIZATION] = "$tokenType $accessToken"
        }
        headers[HttpHeaders.CONTENT_TYPE] = contentType
        
        val response = transportClient[stepikApi, url, headers]
        
        if (response.statusCode / 100 != 2) {
            val message = "Failed query to $url returned the status code ${response.statusCode}"
            logger.warn(message)
            
            if (response.statusCode == StatusCodes.SC_UNAUTHORIZED) {
                throw StepikUnauthorizedException(message)
            } else {
                throw StepikClientException(message)
            }
        }
        
        return response.body
    }
    
    companion object {
        private val logger = LoggerFactory.getLogger(StepikFilesAction::class.java)
    }
}
